complex/cargo/store/src/lib.rs

```diff
@@ -58,34 +141,40 @@ impl fmt::Debug for Store {
 }
 
 impl Store {
-    pub fn new<T>(uri: T) -> Self
+    pub fn new<T>(uri: T) -> Result<Self, store_errors::Error>
     where T: Into<Option<String>> {
-        let uri_string = match &(uri.into()) {
-            &Some(ref u) => u,
-            &None => "",
-        };
+        let uri_string = uri.into().unwrap_or(String::new());
         let mut h = try!(new_connection(&uri_string));
         let c = try!(Conn::connect(&mut h));
         Ok(Store {
-            handle: Arc::new(h),
+            handle: h,
             conn:c,
             uri: uri_string,
         })
     }
 
-    pub fn open(&mut self, uri: String) -> Result<(), mentat::errors::Error> {
-        self.handle = try!(new_connection(&uri));
+    pub fn open<T>(&mut self, uri: T) -> Result<(), store_errors::Error>
+    where T: Into<Option<String>> {
+        let uri_string = uri.into().unwrap_or(String::new());
+        self.handle = try!(new_connection(&uri_string));
         self.conn = try!(Conn::connect(&mut self.handle));
-        self.uri = uri;
+        self.uri = uri_string;
         Ok(())
     }
 
-    pub fn query(&self, query: String) -> Result<QueryResults, mentat::errors::Error> {
-        Ok(self.conn.q_once(&self.handle, &query, None)?)
+    pub fn query<T>(&self, query: &str, inputs: &[&(&String, &T)]) ->  Result<QueryResults, store_errors::Error>
+        where T: ToTypedValue {
+        let mut ee = vec![];
+        for &&(ref arg, ref val) in inputs.iter() {
```

Comment:
You probably want to define `inputs` in terms of `IntoIter`.

Reply:
And probably in terms of `&str` rather than `&String` — `String` derefs into `&str` anyway.